Add Droplet actions

mamercad opened the pull request:
- Refactor module boilerplate into a new class `DigitalOceanCommonModule`
- Add `DIGITALOCEAN_ACCESS_TOKEN` as an additional API token environment variable
- Split Droplet Actions into separate modules
- Add `client_override_options` and `module_override_options` to support experimental or future options
- Remove region validation from argument_specs to facilitate API test beds
- Remove `droplet_action` and turn into `droplet_action_snapshot`
- Add `droplet_action_resize` and `droplet_action_power` and tests
- Add missing Droplet `user_data` parameter
- Fix #44

artis3n commented on .github/workflows/pull-request-integration.yml:
```diff
@@ -44,7 +44,11 @@ jobs:
           - domains_info
           - droplet
-          - droplet_action
+          - droplet_action_power_off
```
The way we use `pull_request_target`, it is expected that the `droplet_action` integration test will fail, because it exists on main but no longer in this branch. This means we can't test integration automatically in the PR, but saving myself a note to run it locally to verify good functionality before ✅

artis3n commented on plugins/modules/droplet_action_power_on.py:
```diff
@@ -0,0 +1,278 @@
+- name: Power on a Droplet
+  digitalocean.cloud.droplet_action_power_on:
+    token: "{{ token }}"
+    state: present
```
What does `state` mean for these individual droplet actions? Perhaps for the power ones alone, we should just have `droplet_action_power` with present/absent state meaning power off or power on. Or custom state values, to account for power on/power off/shutdown.

mamercad replied:
> What does `state` mean for these individual droplet actions? Perhaps for the power ones alone, we should just have `droplet_action_power` with present/absent state meaning power off or power on. Or custom state values, to account for power on/power off/shutdown.

Yes, this is a good suggestion; having a single `digitalocean.cloud.droplet_action.power` with `state` in `{power_on,power_off,shutdown}` is syntactically much more intuitive (and prevents one from having to code or think about what it would mean for `digitalocean.cloud.droplet_action_power_on` with `state=absent` - yes it makes sense that it might be "turn it off" but then why do we have `digitalocean.cloud.droplet_action_power_off`, etc). I'll work on this change.

mamercad replied:
> > What does `state` mean for these individual droplet actions? Perhaps for the power ones alone, we should just have `droplet_action_power` with present/absent state meaning power off or power on. Or custom state values, to account for power on/power off/shutdown.
> 
> Yes, this is a good suggestion; having a single `digitalocean.cloud.droplet_action_power` with `state` in `{power_on,power_off,shutdown}` is syntactically much more intuitive (and prevents one from having to code or think about what it would mean for `digitalocean.cloud.droplet_action_power_on` with `state=absent` - yes it makes sense that it might be "turn it off" but then why do we have `digitalocean.cloud.droplet_action_power_off`, etc). I'll work on this change.

This implementation has been pushed 🚀

artis3n commented on tests/integration/targets/billing_history_info/tasks/main.yml:
```diff
@@ -13,23 +14,24 @@
-    - name: Assert result message
-      ansible.builtin.assert:
-        that: result.msg is string
+    - name: Debug
```
Troubleshooting debug statement, or do you want to keep this in the PR?

artis3n reviewed (commented):
Running these integration tests locally since they won't run in CI until this PR is merged, I have failing tests running `ansible-test integration <module>`:

`droplet_action_power`

![image](https://github.com/digitalocean/ansible-collection/assets/6969296/88fec296-b05c-4fe9-9264-7ab456efc10a)

`droplet_action_resize:

![image](https://github.com/digitalocean/ansible-collection/assets/6969296/093d647c-445c-4dde-8bda-4f890fc087af)

`droplet_action_snapshot`:

![image](https://github.com/digitalocean/ansible-collection/assets/6969296/dc1ce4c4-c205-45f6-a87d-c77ae297b1a4)

artis3n commented:
Would begin troubleshooting them, but I ran the tests through my changes in #53 so I can't push up any commits to this branch

artis3n commented:
@mamercad Because it is impossible to find the review comment in the conversation history, bubbling up: https://github.com/digitalocean/ansible-collection/pull/45#pullrequestreview-1633503747

and https://github.com/orgs/community/discussions/67649

mamercad commented:
> Based on our conversations, it looks like this is the expected result from `droplet_action_power`, so this test looks good now:
> 
> ![image](https://user-images.githubusercontent.com/6969296/268981356-21b4b4de-d44e-4c45-9a90-d0fc829be955.png)
> 
> **Still seeing errors on resize**, I don't think you worked on them yet but since asked for another review here's what I'm seeing:
> 
> `droplet_action_resize`:
> 
> ![image](https://user-images.githubusercontent.com/6969296/268981615-c0d72244-ea19-4158-ae08-067724638a70.png)
> 
> `droplet_action_snapshot`:
> 
> I fixed some typos and pushed them to this branch (references to `droplet_action` instead of migrated `droplet_action_snapshot`). I am getting intermittent passes and failures with this test, so we still need to dig into it.
> 
> Passing:
> 
> ![image](https://user-images.githubusercontent.com/6969296/268999097-d695d333-6923-4114-aebc-df4f1c12395c.png)
> 
> Failing:
> 
> ![image](https://user-images.githubusercontent.com/6969296/268999174-af4db10f-b2c2-436e-826b-e2b51a03d0b3.png)
> 
> (I had 1 set as my PR_NUMBER)

I found and fixed a few things, please try again 🚀

artis3n reviewed (approved):
Integration tests passing locally now!
